packages/gatsby/src/schema/resolvers.ts

```diff
@@ -244,38 +245,52 @@ export const group: GatsbyResolver<
 
   return Object.keys(groupedResults)
     .sort()
-    .reduce((acc: IGatsbyGroupReturnValue<any>, fieldValue: string) => {
+    .reduce((acc: IGatsbyGroupReturnValue<IGatsbyNode>, fieldValue: string) => {
+      const entries = groupedResults[fieldValue] || []
       acc.push({
-        ...paginate(groupedResults[fieldValue], args),
+        ...paginate({ entries, totalCount: entries.length }, args),
         field,
         fieldValue,
       })
       return acc
     }, [])
 }
 
-export function paginate<NodeType>(
-  results: Array<NodeType> = [],
-  { skip = 0, limit }: { skip?: number; limit?: number }
-): IGatsbyConnection<NodeType> {
-  if (results === null) {
-    results = []
+export function paginate(
+  results: IQueryResult,
+  params: { skip?: number; limit?: number; resultOffset?: number }
+): IGatsbyConnection<IGatsbyNode> {
+  const { resultOffset = 0, skip = 0, limit } = params
+  if (resultOffset > skip) {
+    throw new Error("Result offset cannot be greater than `skip` argument")
   }
+  let countOrThunk = results.totalCount
+  const allItems = Array.from(results.entries)
 
-  const count = results.length
-  const items = results.slice(skip, limit && skip + limit)
+  const start = skip - resultOffset
+  const items = allItems.slice(start, limit && start + limit)
```

Comment:
can't this grow pretty big?

Reply:
Well, essentially it happens before this PR - you get 100,000 items in results from fast filters and then slice it here in pagination (e.g. get a slice of 20 items only). But this array contains references to nodes which are already in-memory. Given each reference is ~8 bytes, then slice with 1000 elements will take ~8kb additional memory.

With this PR slicing occurs in the datastore itself. In pagination you get up to 22 items (20 + 1 item before and 1 after). 
Also for lmdb case this `results` will be an iterable and slicing will be lazy as you traverse (in theory - because we do multiple traversals over those results anyways and it might be actualy cheaper to load all 20 items in memory at once).

Reply:
👍 Code is hard to follow on who's slicing where but sweet that lmdb does the slicing for us

Reply:
Yeah, pagination logic with those tricks (like adding additional elements to the left and right) is often puzzling but I've added some tests to demonstrate intended usage. Happy to make it clearer if you have any suggestions 😉